crates/merod/src/cli/config.rs

```diff
@@ -50,38 +159,424 @@ impl ConfigCommand {
+            return Err(eyre!("Empty key provided"));
+        }
+
+        let mut current = doc.as_item_mut();
+
+        for key in &key_parts[..key_parts.len() - 1] {
+            let needs_insert = match current.get_mut(*key) {
+                Some(item) if item.is_table() => false,
+                Some(_) | None => true,
+            };
+            if needs_insert {
+                current[*key] = Item::Table(toml_edit::Table::new());
+            }
+
+            current = current.get_mut(*key).ok_or_else(|| {
+                eyre!(
+                    "Failed to access key '{}' while processing '{}'",
+                    key,
+                    key_parts.join(".")
+                )
+            })?;
+
+            if !current.is_table() {
+                return Err(eyre!(
+                    "Cannot create nested key '{}' - parent '{}' is not a table",
+                    key_parts.join("."),
+                    key
+                ));
+            }
+        }
+
+        let last_key = key_parts[key_parts.len() - 1];
+
+        // Validate type if the key already exists
+        if let Some(existing) = current.get(last_key) {
+            if existing.is_table() && !value.is_inline_table() {
+                return Err(eyre!(
+                    "Cannot set primitive value on existing table '{}'",
+                    key_parts.join(".")
+                ));
+            }
+        }
+
+        current[last_key] = Item::Value(value);
+        Ok(())
+    }
+
+    fn get_schema() -> HashMap<String, ConfigSchema> {
+        let mut schema = HashMap::new();
+
+        // Network configuration
+        let mut network = HashMap::new();
+
+        // Discovery config
+        let mut discovery = HashMap::new();
+        discovery.insert(
+            "mdns".to_owned(),
+            ConfigSchema {
+                description: Some("Enable mDNS discovery".to_owned()),
+                type_info: ConfigType::Boolean,
+                default: Some(SchemaValue::Boolean(true)),
+            },
+        );
+        discovery.insert(
+            "advertise_address".to_owned(),
+            ConfigSchema {
+                description: Some("Advertise observed address".to_owned()),
+                type_info: ConfigType::Boolean,
+                default: Some(SchemaValue::Boolean(false)),
+            },
+        );
+
+        network.insert(
+            "discovery".to_owned(),
+            ConfigSchema {
+                description: Some("Discovery configuration".to_owned()),
+                type_info: ConfigType::Object(Box::new(discovery)),
+                default: None,
+            },
+        );
+
+        schema.insert(
+            "network".to_owned(),
+            ConfigSchema {
+                description: Some("Network configuration".to_owned()),
+                type_info: ConfigType::Object(Box::new(network)),
+                default: None,
+            },
+        );
+
+        schema
+    }
```

Comment:
as previously mentioned, we shouldn't do this manually. The code should be the source of truth, otherwise there's an implicit maintainance burden here. All changes to the actual config structure will have to remember to update these, which makes it not worthwhile.

[`JSONSchema`](https://docs.rs/schemars/latest/schemars/) solves all of this, and this PR is not considered ready or complete without it, unfortunately.

Unless we choose to make that a separate effort, completely remove hinting and only deliver editing and general printing for this one.

Reply:
Alright, after reading more about the JSONSchema, I think I understand it better now. I'll push an update before the end of the day.